qiskit/dagcircuit/dagdependency_v2.py

```diff
@@ -0,0 +1,725 @@
+from qiskit.circuit.controlflow import condition_resources
+from qiskit.circuit.bit import Bit
+from qiskit.dagcircuit.dagnode import DAGOpNode
+from qiskit.dagcircuit.exceptions import DAGDependencyError
+from qiskit.circuit.commutation_checker import CommutationChecker
+
+
+BitLocations = namedtuple("BitLocations", ("index", "registers"))
+
+
+class DAGDependencyV2:
+    """Object to represent a quantum circuit as a Directed Acyclic Graph (DAG)
+    via operation dependencies (i.e. lack of commutation).
+
+    The nodes in the graph are operations represented by quantum gates.
+    The edges correspond to non-commutation between two operations
+    (i.e. a dependency). A directed edge from node A to node B means that
+    operation A does not commute with operation B.
+    The object's methods allow circuits to be constructed.
+
+    **Example:**
+
+    Bell circuit with no measurement.
+
+    .. parsed-literal::
+
+              ┌───┐
+        qr_0: ┤ H ├──■──
+              └───┘┌─┴─┐
+        qr_1: ─────┤ X ├
+                   └───┘
+
+    The dependency DAG for the above circuit is represented by two nodes.
+    The first one corresponds to Hadamard gate, the second one to the CNOT gate
+    as the gates do not commute there is an edge between the two nodes.
+
+    **Reference:**
+
+    [1] Iten, R., Moyard, R., Metger, T., Sutter, D. and Woerner, S., 2020.
+    Exact and practical pattern matching for quantum circuit optimization.
+    `arXiv:1909.05270 <https://arxiv.org/abs/1909.05270>`_
+
+    """
+
+    def __init__(self):
+        """
+        Create an empty DAGDependencyV2.
+        """
+        # Circuit name
+        self.name = None
+
+        # Circuit metadata
+        self.metadata = {}
+
+        # Cache of dag op node sort keys
+        self._key_cache = {}
+
+        # Directed multigraph whose nodes are operations(gates) and edges
+        # represent non-commutativity between two gates.
+        self._multi_graph = rx.PyDAG()
+
+        # Map of qreg/creg name to Register object.
+        self.qregs = OrderedDict()
+        self.cregs = OrderedDict()
+
+        # List of Qubit/Clbit wires that the DAG acts on.
+        self.qubits: List[Qubit] = []
+        self.clbits: List[Clbit] = []
+
+        # Dictionary mapping of Qubit and Clbit instances to a tuple comprised of
+        # 0) corresponding index in dag.{qubits,clbits} and
+        # 1) a list of Register-int pairs for each Register containing the Bit and
+        # its index within that register.
+        self._qubit_indices: Dict[Qubit, BitLocations] = {}
+        self._clbit_indices: Dict[Clbit, BitLocations] = {}
+
+        self._global_phase = 0
+        self._calibrations = defaultdict(dict)
+
+        # Map of number of each kind of op, keyed on op name
+        self._op_names = {}
+
+        self.duration = None
+        self.unit = "dt"
+
+        self.comm_checker = CommutationChecker()
+
+        # Map of node to node index
+        self.node_map = {}
```

Comment:
Somehow I completely missed that `_node_id` was used throughout DAGCircuit. I thought it was only used minimally, so I didn't use it in DAGDependencyV2, but should have. Since it would also need to be used in the template matching code and the underscore implies some level of privacy, we'd probably want to add an `@property` for `node_id` to the DAGDependencyV2 code.